Allow interface resolveType functions to resolve to child interfaces

yaacovCR opened the pull request:
= The child interfaces must eventually resolve to a runtime object type.

implements: #3253

yaacovCR committed:
Allow interface resolveType functions to resolve to child interfaces

= The child interfaces must eventually resolve to a runtime object type.
= Interface cycles raise a runtime error.

saihaj commented on src/execution/execute.ts:
```diff
@@ -826,32 +825,73 @@ function completeAbstractValue(
   );
 }
 
+function resolveType(
```
I would suggest making it one argument that is an object

yaacovCR replied:
Hmmm, I think most of the code in this file does not have named arguments, see for example nearby `ensureValidRuntimeType` and many others.

The public API has been moving in the direction of named parameters, probably for good reason. The functions in this file by contrast are not exported and only used a few times, so the benefits of named parameters that leap to mind (any order, can save a set of arguments and pass them around) do not seem as straightforward.

But I defer to the group, thanks for the eyeballs.

yaacovCR replied:
Actually, I think you're right, in this case we have a recursive function that is passing the exact same set of arguments save one, they should be combined.

yaacovCR replied:
Hmmm, I'm looking at it, and it seems like similar refactoring could occur in so many places. I'm going to leave this as is for now.

IvanGoncharov commented on src/execution/execute.ts:
```diff
@@ -826,32 +825,73 @@ function completeAbstractValue(
+  );
+}
+
 function ensureValidRuntimeType(
```
this function is called "ensure" (and previously it just validated type) but now actually recursively resolves type. Please change its name?

yaacovCR replied:
Sure => what do you suggest? I would like something that implies ensure + recursion as necessary. I guess I thought ensure was good enough, but if you have some candidates that you would prefer, that would be fine.

yaacovCR replied:
I gave it a go => I was more explicit that we start off with the type name and changed the function to `deriveRuntimeType`.

6c63b4f2

IvanGoncharov commented on src/execution/execute.ts:
```diff
@@ -826,32 +825,73 @@ function completeAbstractValue(
+  );
+
+  if (isPromise(possibleRuntimeType)) {
+    return possibleRuntimeType.then((resolvedPossibleRuntimeType) =>
```
Do you really need to add `resolveType` function to implement this feature or it's just refactoring mixed in?

The problem here is that adding one more `then` looks unnecessary.
Every `then` adds a trip to the microtasks queue even if the promise is already resolved and that significantly affects performance.
Because of that code, some time can look repetitive but it is done on purpose to minimize the number of `then` and improve performance.

yaacovCR replied:
@IvanGoncharov, I totally agree with you, but that leaves me a bit puzzled.

This code actually is repetitive; I switch on whether the type resolver for the given abstract type (or the default type resolver) returned a promise, and only call `.then` when it is a promise. For exactly the reason you mention above, to not use then unnecessarily.

For that reason, I did need the `resolveType` function -- or, at least I thought I did.

Maybe you could be more specific about where you think the extra trip to the microtask queue is happening?

yaacovCR replied:
feel free to directly modify this PR if you have something else in mind.

yaacovCR replied:
summary of offline conversation:

> I am checking if we have a promise appropriately at each resolveType call, but because I refactored a recursive function that returned a PromiseOrValue, when there is no extra recursive steps, I ended up calling `then` twice: once for the only call to the possibly async type resolver, but also once on the entire recursive function.
>
> The recursion can be refactored (actually much closer to the original) so that if there is no need for recursion, the recursive part is not called.

IvanGoncharov commented on src/execution/execute.ts:
```diff
@@ -866,29 +906,62 @@ function ensureValidRuntimeType(
   if (!exeContext.schema.isSubType(returnType, runtimeType)) {
     throw new GraphQLError(
-      `Runtime Object type "${runtimeType.name}" is not a possible type for "${returnType.name}".`,
+      `Abstract type resolution for "${returnType.name}" for field "${info.parentType.name}.${info.fieldName}" failed. ` +
```
You still need to check if the intermediate interface is a sub-type of `returnType` and currently you bypassing this check.
Also please add a test for it.

yaacovCR replied:
There is a test a few lines above for the intermediate interface.

How is the test being bypassed?

yaacovCR replied:
ah, you meant that i am checking against `returnType` but not the current abstract type. fixed, new test added.

We should get an error if we go from Named => Pet => Person, even though Person is Named, it's not a Pet.

df8f15c7

yaacovCR replied:
7ceffbe8

yaacovCR replied:
7ceffbe8

yaacovCR replied:
1e2988e

yaacovCR committed:
change function name

yaacovCR commented on src/execution/__tests__/abstract-test.ts:
```diff
@@ -271,7 +271,7 @@ describe('Execute: Handles execution of abstract types', () => {
         {
           message:
-            'Abstract type "Pet" must resolve to an Object type at runtime for field "Query.pet". Either the "Pet" type should provide a "resolveType" function or each possible type should provide an "isTypeOf" function.',
+            'Abstract type resolution for "Pet" for field "Query.pet" failed. Encountered abstract type "Pet" must resolve to an Object or Interface type at runtime. Either the "Pet" type should provide a "resolveType" function or each possible type should provide an "isTypeOf" function.',
```
The error messages have to be changed (in my opinion) because there is now a multi-step process for deriving the runtime type, and we want to be as helpful as possible by pointing out what type we started with and where we failed. It's not a separate change => it's just part and parcel of what we're doing here, it wouldn't make sense on its own, it would just be repetitive unless we have the multi-step process.

IvanGoncharov replied:
> and we want to be as helpful as possible by pointing out what type we started with

Please note that this is not a **client** error, the client can't trigger that error unless there is a bug in type resolvers. So for a client having this info is useless and it should matter only for people working on a schema.

Moreover, for clients, this detail is already accessible through the `path` and `locations`.
GraphiQL will highlight the field you get an error on and you will see its type in a tooltip.

I'm not sure having it explicitly in the message just for schema developers justifies an additional complexity of passing additional arguments just for error reporting.

In general, type resolvers should be independent of each other same as field resolvers. So it shouldn't matter what was previous type resolver since it's doesn't affect the resolver's argument.
If the schema author really wants to debug it when he can easily use the available type name and field name to quickly find initial type in a chain.

yaacovCR replied:
From my survey of the repo and at GraphQL-Tools, there are many schema developers out there who get confused by cryptic error messages from this (and many other!) packages. I am in favor of keeping more details in the error message, perhaps I can refactor per @saihaj  suggestion to just pass around one object that bundles some of the constant parameters?

yaacovCR replied:
Actually, I think it just makes sense to go with the shorter messages. We can always change it later.

yaacovCR commented on src/execution/__tests__/abstract-test.ts:
```diff
@@ -640,4 +640,106 @@ describe('Execute: Handles execution of abstract types', () => {
+
+    // FIXME: workaround since we can't inject resolveType into SDL
+    // @ts-expect-error
+    assertInterfaceType(schema.getType('Pet')).resolveType = () => [];
```
This is a child interface rather than the initial abstract type, even though I know the same function is being used, I am not making assumptions, and adding a test for that as well.

IvanGoncharov replied:
In that case, please add the same test for intermediate interfaces.

IvanGoncharov replied:
got confused, now I see `Pet` is intermediate interface
👍

yaacovCR committed:
use variables for abstract types in tests

yaacovCR commented on src/execution/__tests__/abstract-test.ts:
```diff
@@ -640,4 +640,106 @@ describe('Execute: Handles execution of abstract types', () => {
+
+    // FIXME: workaround since we can't inject resolveType into SDL
+    assertInterfaceType(schema.getType('Named')).resolveType = () => 'Pet';
+    assertInterfaceType(schema.getType('Pet')).resolveType = () => undefined;
```
sure thing  46ae4b19

yaacovCR committed:
fix chain and add test

yaacovCR committed:
standardize order

IvanGoncharov commented on src/execution/execute.ts:
```diff
@@ -866,29 +906,62 @@ function ensureValidRuntimeType(
+
+  if (encounteredTypeNames.has(runtimeTypeName)) {
+    throw new GraphQLError(
+      `Abstract type resolution for "${returnType.name}" for field "${info.parentType.name}.${info.fieldName}" failed. ` +
```
We can't have cycles, they were possible only because previously you didn't do a subtype check (previous review comment).
Schema specifically validated to not have interface cycles, see schema validation code.
I already fixed this as part of previous comment will push on top of your PR.

IvanGoncharov replied:
Pushed, please review.

yaacovCR replied:
Some schemas are not validated i.e. "assumeValid" is true, and so I imagined we might still need the cycle check. I could be off-base.

yaacovCR replied:
i added a test for an unvalidated schema to show why this is still necessary.

IvanGoncharov replied:
As the comment says it's up to a person who sets the `assumeValid` flag is a guarantee that schema is valid.
https://github.com/graphql/graphql-js/blob/7532a5c8eb4001b726a2b45c8760bb71a6e10f2e/src/type/schema.ts#L416-L418

We can't assume something is broken when a person explicitly states it's valid.
Especially in performance-sensitive code.

yaacovCR replied:
I am basing off how unvalidated fragment cycles are not allowed to infinite loop.

IvanGoncharov committed:
remove unneeded check for cycles

yaacovCR commented:
hi @IvanGoncharov just checking on the status here, no rush, just curious!

IvanGoncharov committed:
review change

yaacovCR commented:
This is ready for re-review, thanks.

yaacovCR commented:
Hey @IvanGoncharov  looks like we both solved the extra then but I overwrote your changes with mine. I am away from computer until about 15 hours from now, sorry about that and feel free to restore yours. Thanks for your help advancing this

yaacovCR commented:
> Hey @IvanGoncharov looks like we both solved the extra then but I overwrote your changes with mine. I am away from computer until about 15 hours from now, sorry about that and feel free to restore yours. Thanks for your help advancing this

I was able to fix before work yay

yaacovCR commented on src/execution/execute.ts:
```diff
@@ -795,63 +797,63 @@ function completeLeafValue(
 
-function ensureValidRuntimeType(
-  runtimeTypeName: unknown,
+function resolveRuntimeTypeName(
```
This function now essentially completes, suggest renaming to completeAbstractValueImpl or something else.

yaacovCR replied:
i pushed a change renaming it

yaacovCR commented:
good news! i rebased on main and after #3624 , the report uploads correctly!

yaacovCR committed:
change inner recursive function to completeAbstractValueImpl

yaacovCR committed:
re-align error message as possible to original

yaacovCR committed:
rename runtimeType(Name) to possibleRuntimeType(Name)

because it could not be the runtime type(name), could be an intermediate interface

I considered just renaming it to resolvedTypeName, but that's confusing, because the variable could hold a promise, and resolvedResolvedTypeName overloads "resolve"

IvanGoncharov reviewed (changes requested):
This PR still has some unfinished discussions.
But I discovered one major roadblock, it requires spec PR, because otherwise, it contradicts spec:
https://spec.graphql.org/draft/#sec-Value-Completion.Resolving-Abstract-Types

I suggest opening spec PR that changes the algorithm to be recursive.
\+ I will vote for fast-track it to stage2 if you present it on the next WG.

yaacovCR commented:
Two things:

### First, what other discussions are pending?

The two pending issues I had seen were the content of the error messages and whether cycles should be allowed, and I have changed this PR to correspond to your suggestions. Are you not sure of the optimal behavior with respect to one or both of these issues? Do you feel that you need a certain amount of time to think about these issues, or do these issues need feedback from the graphql-js-wg or the main graphql-wg? I will address your comment in a second, but we can work on these issues in parallel, as although async discussions have their drawback, they do allow us to open up multiple threads and make progress in parallel.

### Second, I think this PR does not contradict spec, as the spec says we can resolve by whatever means we want.

In particular, consider the evolution of this PR. A previous version had a separate resolve method corresponding to the ResolveAbstractType step you have linked. But:
1. that method returned a value or promise, and
2. refactoring any functionality returning a promise or value into a separate function forces the calling function to perform an additional isPromise check and `.then` call
3. `graphql-js` has the aspirational policy of not only not entering the task queue unnecessarily, but of never adding an additional job to job queue (see https://blog.greenroots.info/task-queue-and-job-queue-deep-dive-into-javascript-event-loop-model and https://developer.mozilla.org/en-US/docs/Web/API/HTML_DOM_API/Microtask_guide/In_depth#run_javascript_run)

And so we had to mix the separate `resolveAbstractType` method with abstract value completion using recursive `completeAbstractValue` and `completeAbstractValueImpl` inner functions and an incomplete `resolveAbstractType` portion linking between the two.

So, this change does not contradict the spec, but the performance (?) constraints of `graphql-js` have causes us to make a change that makes the one-to-one match between the implementation and the spec more difficult to grasp.

[Perhaps there are 3 tenets of `graphql-js`, [not 2!](https://github.com/graphql/graphql-js/issues/3519#issuecomment-1129850480): (1) a reference implementation, (2) a production-ready JavaScript graphql pipeline, (3) a fully-fleshed out exercise in avoiding any extra jobs in the job queue. And perhaps there is some sort of inchoate rule where for any given issue, only two of the three `graphql-js` tenets may be upheld!]

Bottom line, I don't think we should necessarily force all complying implementations to implement this feature in JavaScript. I think it's a useful feature, but I don't know that in every language it should be mandated.

However, I am fine:

1. working on a spec PR
2. presenting it at the next WG
3. adding that my own opinion is that we should implement this in `graphql-js` without adopting the spec PR, and then
4. accepting whatever happens

Although please see above -- let's work out the other issues in the meantime. I am leaving this as open, please follow-up when possible.

yaacovCR commented:
See:
https://github.com/graphql/graphql-spec/pull/960
https://github.com/graphql/graphql-wg/pull/1016

leebyron commented on src/execution/execute.ts:
```diff
@@ -790,106 +791,128 @@ function completeLeafValue(
   }
 
-  if (!isObjectType(runtimeType)) {
+  if (isInterfaceType(possibleRuntimeType)) {
```
what about union types?

yaacovCR replied:
Yes, if unions ever implement interfaces, they could potentially be a link in the chain!

leebyron commented on src/execution/execute.ts:
```diff
@@ -790,106 +791,128 @@ function completeLeafValue(
+  if (possibleRuntimeTypeName == null) {
     throw new GraphQLError(
-      `Abstract type "${returnType.name}" must resolve to an Object type at runtime for field "${info.parentType.name}.${info.fieldName}". Either the "${returnType.name}" type should provide a "resolveType" function or each possible type should provide an "isTypeOf" function.`,
+      `Abstract type "${abstractType.name}" must resolve to an Object type or an intermediate Interface type at runtime for field "${info.parentType.name}.${info.fieldName}". Either the "${abstractType.name}" type should provide a "resolveType" function or each possible type should provide an "isTypeOf" function.`,
```
What do you think about only altering this error message in the case that an intermediate Interface type actually exists as a potential case?

Looking at the unit test impact, my sense is the primary downside of this change is the additional complication to this error message that makes it slightly harder to understand.

Ideally for the large majority common case where there is no intermediate interface, we don't burden the error message

yaacovCR replied:
👍  Tailoring the error message to the interface at hand seems pretty reasonable.

yaacovCR commented:
Closing this, can be used as base if anyone wants to champion the associated RFC
